### ISSUE TITLE
Create NixTeX

### ISSUE THREAD

**Issue by edolstra** [NixOS/nixpkgs#287]
TeX Live is an insane piece of bloatware: an installation of the texLiveFull attribute in Nixpkgs takes > 3.1 GB. By contrast, teTeX was "only" 261 MB. There are even quite a few packages in Nixpkgs that unwisely have a dependency on TeX Live, thereby adding 3 GB to their (build) closure.

We should be able to do better in Nixpkgs by properly componentizing TeX Live. Ideally, every CTAN package would be its own Nixpkgs package, with proper dependencies. Then other packages can have finegrained dependencies on those TeX packages that they need. And of course there should be some wrapper packages like texLiveFull to allow easy installation of all or some reasonable subset of the distribution.

**Comment by matthiasbeyer** [comment]
This issue is also an really old one and it seems that there is no progress here anymore. Does anyone care closing this?

**Comment by vcunat** [comment]
@matthiasbeyer: the issue is more of a TODO item (there's been practically no work on this yet AFAIK).

**Comment by vcunat** [comment]
I'm working on this currently. Overview of the approach and currently largest stumbling blocks: http://tug.org/pipermail/tldistro/2015q3/000253.html

WIP is https://github.com/NixOS/nixpkgs/compare/master...vcunat:p/tex-space-stdenv?expand=1

**Comment by edolstra** [comment]
FWIW, here are the Nix expressions I made for a few texlive components: http://pastebin.com/AdZPXTFQ

The main reason for separating out programs like xdvi was to prevent the rest of TeX from depending on X11, ghostscript etc.

**Comment by vcunat** [comment]
First I want to make package composing work, as the many-gigabyte texmf seems the largest problem.

I was considering whether to just reduce runtime deps (multiple outputs) or split properly like what you started. My last plan was to keep building the dependency-light stuff in one build and one output, and separate just a few individual heavy packages like xdvik or xetex. The texlive docs claim that many parts won't be straightforward to build separately...

**Comment by vcunat** [comment]
@edolstra: [CTAN won't mirror](http://tug.org/pipermail/tldistro/2015q3/000255.html) a stable version of individual tarballs; will it be acceptable for our hashedMirror to cover at least the most important parts? I expect ~100-200 MB (shared for all platforms, stdenvs, etc.). For the rest, I expect this slow URL will suffice: ftp://tug.ctan.org/pub/tex/historic/systems/texlive/2015/tlnet-final/ (at least for me it's slow).

**Comment by Mathnerd314** [comment]
Another option would be to use a Fedora SRPM as the source, e.g. the file in https://kojipkgs.fedoraproject.org//packages/texlive/2015/4.20150728_r37987.fc24/src/ (it contains a snapshot of the tar.zx's). Although Nix doesn't support [multiple-output fixed-output derivations](https://github.com/NixOS/nix/blob/master/src/libexpr/primops.cc#L606).

**Comment by vcunat** [comment]
I wanted to avoid having to download a 1.8 GB tarball even if you only want to install some small parts (hundreds of MB).

**Comment by vcunat** [comment]
So, after lots of work, it seems very good to me, at least for standard LaTeX stuff. See [basic user docs](https://github.com/vcunat/nixpkgs/blob/p/tex-space-stdenv/pkgs/tools/typesetting/tex/texlive-new/default.nix).

It would be nice if you tested it on your use cases, especially if you use something less usual than LaTeX with some extra packages. My branch is based on a binary-cached channel, so you should get most things fast.

Note that ATM the standard LaTeX env (`texlive.combined.scheme-basic`) uses **~70 MB** (!) of texlive packages + perl and bash, and it depends on no heavy stuff, i.e. no X, even during build time.

**Comment by abbradar** [comment]
I've tried building `texlive.combined.scheme-small` on your branch; it failed with invalid md5 sum for `xetex` [1] and then failed with this:
```
building path(s) ‘/nix/store/vas0mlfzh32la859qab8qls0132j49d3-texlive-combined-2015’
created 479 symlinks in user environment
cp: cannot stat '/nix/store/vas0mlfzh32la859qab8qls0132j49d3-texlive-combined-2015/share/texmf/scripts/texlive/fmtutil.pl': No such file or directory
builder for ‘/nix/store/n8pg0wnsfj6ysjk80p1mblywrypfbvi5-texlive-combined-2015.drv’ failed with exit code 1
```
I guess it should have been in `tetex` but I don't know much about structure of LaTeX packages...

1: https://github.com/abbradar/nixpkgs/commit/c0847c8e5d515b97cc65c30024eab5cc00ab0e8b

**Comment by vcunat** [comment]
Ah, I forgot to fix that TODO (read comments above your changed lines for details). For a quick work around, run ```nix-prefetch-url --type md5 http://ftp.cvut.cz/tex-archive/systems/texlive/tlnet/archive/tetex.tar.xz`` instead of that patch of yours.

**Comment by abbradar** [comment]
I've tried to build some documents of mine with the new `texlive`. Impressive work, I should say! My personal wishlist (overlapping with your TODO) would be:

1. Fix `xelatex` problems with fonts
2. Package `latexmk` properly (I suppose this would involve just a `makeWrapper` to `/bin` with a proper `PATH`)
3. Package `biber`

**Comment by abbradar** [comment]
Hm, we already have `biber` packaged separately, so I suppose you don't need to redo the work. Maybe `inherit`ing it into `texlive` would suffice.

**Comment by vcunat** [comment]
Any TeX-extending packages outside `texlive` likely won't work together with it without changes. Thanks for the feedback! I'm still thinking of doing the `pkgFilter` thing in a different way from user point of view, but maybe I'll just leave it. I do want to have this in 15.09 in some form.

**Comment by abbradar** [comment]
Well, fixing `biber` would be low-priority because one can install and use it separately -- having it in `PATH` should suffice. Maybe making `pkgFilter` a list of outputs instead of a function would suffice? This should be simple and look better.

**Comment by vcunat** [comment]
Yeah, well *adding* a simpler alternative is easy to do. I rather felt I might want to do things it can't express :-)

**Comment by edolstra** [comment]
Hm, isn't it a bit late for such major changes in 15.09?

**Comment by vcunat** [comment]
- I don't consider it too major, as very few packages depend on it (even transitively).
- I'm leaving the old `texLive` infrastructure in for the time being.

**Comment by vcunat** [comment]
But personally, I care very little. I can use `texlive` from unstable easily.

**Comment by abbradar** [comment]
@vcunat I can't think of any use case of having a function but I don't really know what's going on here -- I might easily have missed something.

**Comment by abbradar** [comment]
Actually I found one usecase in your manual just now, so consider my thoughts above invalid:
```
      pkgFilter = pkg:
        pkg.tlType == "run" || pkg.tlType == "bin" || pkg.pname == "cm-super";
```

**Comment by vcunat** [comment]
Currently I force-filter things like `metapost.bin`, leaving only the core binary package, as I don't need them and they bring quite large closures (X, cairo). Such binaries are included by upstream even into relatively basic schemes/collections. I reduced this bloat a little already by removing some of them from "collection-basic" https://github.com/vcunat/nixpkgs/blob/p/tex-space-stdenv/pkgs/tools/typesetting/tex/texlive-new/default.nix#L75, but there are others in slightly larger collections and I don't (yet) like to diverge from upstream TL too much.
```nix
{  pkgFilter = pkg: pkg.tlType == "run"
    || pkg.pname == "core";
/*...*/ }
```

**Comment by vcunat** [comment]
So all scripts linked by upstream (e.g. `latexmk`) are now linked as well, and their shebangs are patched during `combine` even if they need python or ruby (found 4+1 such pkgs for now).

**Comment by abbradar** [comment]
I tried to compile a simple document trying to fix XeLaTeX fonts problem and ran into this:
```
 /tmp  ~/nixpkgs/result/bin/xelatex test.tex 
This is XeTeX, Version 3.14159265-2.6-0.99992 (Web2C 2015/NixOS.org) (preloaded format=xetex)
 restricted \write18 enabled.

kpathsea: Running mktexfmt xetex.fmt
fmtutil [ERROR]: missing command; try fmtutil --help if you need it.
warning: kpathsea: mktexfmt output `fmtutil: fmtutil is using the following fmtutil.cnf files (in precedence order):
fmtutil:   /nix/store/x0v9p0qp1d8xcqmwgdvgcnx3kckbn9x2-texlive-combined-2015/share/texmf/web2c/fmtutil.cnf
fmtutil: fmtutil is using the following fmtutil.cnf file for writing changes:
fmtutil:   /home/abbradar/.texlive2014/texmf-config/web2c/fmtutil.cnf' instead of a filename.
I can't find the format file `xetex.fmt'!
 /tmp  
```
That wasn't the problem before. Any ideas?

**Comment by abbradar** [comment]
Looks like this is a combination of bugs. Yet another log output, after https://github.com/vcunat/nixpkgs/commit/aacafeaefbc937a559bb409308f0854969ac6be8:
```
This is XeTeX, Version 3.14159265-2.6-0.99992 (Web2C 2015/NixOS.org) (preloaded format=xetex)
 restricted \write18 enabled.

kpathsea: Running mktexfmt xetex.fmt
mktexfmt: mktexfmt is using the following fmtutil.cnf files (in precedence order):
mktexfmt:   /nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/web2c/fmtutil.cnf
mktexfmt: mktexfmt is using the following fmtutil.cnf file for writing changes:
mktexfmt:   /home/shlomo/.texlive2014/texmf-config/web2c/fmtutil.cnf
mktexfmt: running `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' ...
This is XeTeX, Version 3.14159265-2.6-0.99992 (Web2C 2015/NixOS.org) (INITEX)
 restricted \write18 enabled.
entering extended mode

(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/plain/config/xetex.ini

(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/generic/xetexconfig/unicode-letters.tex loading Unicode properties
and character classes)
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/plain/etex/etex.src
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/plain/base/plain.tex Preloading the plain format: codes, registers,
parameters, fonts, more fonts, macros, math definitions, output routines,
hyphenation
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/generic/hyphen/hyphen.tex [skipping from \patterns to end-of-file...]))
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/plain/etex/etexdefs.lib Skipping module "grouptypes";
Loading module "interactionmodes"; Skipping module "nodetypes";
Skipping module "iftypes";)
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/generic/config/language.def
(/nix/store/184z43yx4bp8083lyrd9wd66zml14aqv-texlive-combined-2015/share/texmf/
tex/generic/hyphen/hyphen.tex)
! I can't find file `dehypht-x-2014-05-21.tex'.
\addlanguage ...me =#5 \uselanguage {#1}\input #2 
                                                  \if *#3*\else \input #3 \f...
l.32 ...4-05-21}{dehypht-x-2014-05-21.tex}{}{2}{2}
                                                  
(Press Enter to retry, or Control-D to exit)
Please type another input file name: 
! Emergency stop.
\addlanguage ...me =#5 \uselanguage {#1}\input #2 
                                                  \if *#3*\else \input #3 \f...
l.32 ...4-05-21}{dehypht-x-2014-05-21.tex}{}{2}{2}
                                                  
No pages of output.
Transcript written on xetex.log.
mktexfmt [ERROR]: call to `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' returned error code 1
mktexfmt [ERROR]: `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' failed
mktexfmt [INFO]: Disabled formats: 10
mktexfmt [INFO]: Not selected formats: 45
mktexfmt [INFO]: Failure during builds: 1
I can't find the format file `xetex.fmt'!
```
I've also spotted this in the build log:
```
Wrapping './bin/xelatex'head: cannot open 'xetex' for reading: No such file or directory

Wrapping './bin/xelollipop'head: cannot open 'xetex' for reading: No such file or directory


```

**Comment by vcunat** [comment]
The mktexfmt/fmtutil bug should be fine now (again). I partially forgot this case where argv0 matters. 

I'll have a look at what you get in build log, presumably when installing xetex (I didn't get such errors for my configs).

**Comment by vcunat** [comment]
Yes, it was another manifestation of the same bug, so it should be OK now.

I'm still planning to at least make all non-binary packages fixed-output, so they will be shared regardless of changing platform, stdenv, etc. IMO, that will be useful on Hydra and maybe for some users as well.

**Comment by abbradar** [comment]
I'm getting the second log again now. Potentially interesting parts are:
```
fmtutil [ERROR]: call to `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' returned error code 1
```
in build log, and
```
(/nix/store/bj2mf43q3z33rz4xrxvclbjbk303s00i-texlive-combined-2015/share/texmf/
tex/generic/hyphen/hyphen.tex)
! I can't find file `dehypht-x-2014-05-21.tex'.
\addlanguage ...me =#5 \uselanguage {#1}\input #2 
                                                  \if *#3*\else \input #3 \f...
l.32 ...4-05-21}{dehypht-x-2014-05-21.tex}{}{2}{2}
```
and
```
mktexfmt [ERROR]: call to `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' returned error code 1
mktexfmt [ERROR]: `xetex -ini  -jobname=xetex -progname=xetex -etex xetex.ini' failed
mktexfmt [INFO]: Disabled formats: 10
mktexfmt [INFO]: Not selected formats: 45
mktexfmt [INFO]: Failure during builds: 1
I can't find the format file `xetex.fmt'!
```
in the compiling log.

**Comment by abbradar** [comment]
I have some progress on this -- I've added `dehyph-exptl` package (possible dependency of `xetex`?). Now the error is different:
```
(/nix/store/whgnk3fvsf2gbqfkdyj3bm264x5zmjz1-texlive-combined-2015/share/texmf/
tex/generic/hyph-utf8/patterns/tex/hyph-grc.tex))
! I can't find file `ibyhyph.tex'.
\addlanguage ...me =#5 \uselanguage {#1}\input #2 
                                                  \if *#3*\else \input #3 \f...
l.40 \addlanguage{ibycus}{ibyhyph.tex}{}{2}{2}
                                              
(Press Enter to retry, or Control-D to exit)
Please type another input file name: x
```
There is a mention of that file at http://www.ctan.org/tex-archive/language/greek/ibycus-babel/, but we don't have it in our corresponding package!

**Comment by vcunat** [comment]
That's unrelated: it looks like my hyphenation hack doesn't work for some combinations of packages. This particular file is in `ancient-hyphengreek`, but that's not the point, as we need to reduce the hyphenation list to only those that are installed.

I'll think again on how to approach this. Can you share what set of texlive packages you use so I can reproduce it?

**Comment by abbradar** [comment]
Minimal set: `scheme-basic xetex dehyph-exptl`. See the test command above (I don't really know what it does but from other logs it should finish correctly for `xelatex` to work). Alternatively you can try to compile this:
```
\documentclass{article}

\begin{document}
Lorem ipsum...
\end{document}
```

**Comment by abbradar** [comment]
With `hyphen-ancientgreek` things look even more strange:
```
 /tmp  ~/nixpkgs/result/bin/latex test.tex  
This is pdfTeX, Version 3.14159265-2.6-1.40.16 (TeX Live 2015/NixOS.org) (preloaded format=latex)
 restricted \write18 enabled.
entering extended mode
(./test.tex
LaTeX2e <2014/05/01>
Babel <3.9l> and hyphenation patterns for 3 languages loaded.

(/nix/store/jhc9l54b0n14pjs0z1r0df9pxzqygwk1-texlive-combined-2015/share/texmf/
tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class

(/nix/store/jhc9l54b0n14pjs0z1r0df9pxzqygwk1-texlive-combined-2015/share/texmf/
tex/latex/base/size10.clo)) (./test.aux) [1] (./test.aux) )
Output written on test.dvi (1 page, 232 bytes).
Transcript written on test.log.
 /tmp  ~/nixpkgs/result/bin/xelatex test.tex  
This is XeTeX, Version 3.14159265-2.6-0.99992 (Web2C 2015/NixOS.org) (preloaded format=xetex)
 restricted \write18 enabled.
entering extended mode
(./test.tex
! Undefined control sequence.
l.1 \documentclass
                  {article}
?  
```

**Comment by abbradar** [comment]
Argh! For `xetex` we need wrapper to pass `argv[0]` for `xelatex` to work properly!

**Comment by vcunat** [comment]
The argv0 problem should be gone now.

**Comment by abbradar** [comment]
Looks like all `xelatex` problems are gone now; I was able to compile this:
```
\documentclass{article}

\usepackage{fontspec}
\setmainfont{Source Code Pro}

\begin{document}
Lorem ipsum...
\end{document}
```
having `Source Code Pro` installed in NixOS. Thank you very much for help!

**Comment by vcunat** [comment]
Thanks for testing. It's good to catch most errors while this is fresh in my mind.

The hyphenation problem disappeared then?

**Comment by abbradar** [comment]
No, we still need to specify needed packages for now (`scheme-small xetex dehyph-exptl hyphen-ancientgreek`), but at least there is a workaround ^_^. I'll try to compile something complex now...

**Comment by abbradar** [comment]
Aargh, `argv0` strikes again -- now `latexmk` doesn't start and after a long period of silence stops with `too long argument list` -- there's an infinite recursion somewhere, I think...

**Comment by vcunat** [comment]
Yes, I'm sorry. There was an interaction with my code a few lines below. I was being a bit hasty when fixing the argv0 problem. Now `latexmk` runs fine again for me.

**Comment by vcunat** [comment]
Note: there's a rebuild of packages due to making them fixed-output in 951f4606.

**Comment by abbradar** [comment]
I was able to build my bachelor's thesis using this (with a fairly big amount of packages and "external" `pkgs.biber`), `xelatex` even correctly found and used an OTF font from another package (`lm-math`). Thank you again very much for all this work!

To keep track of broken/improvable things we've discovered so far:

1. Fix `luatex` (I'm looking into this a little, no luck so far)
2. Integrate `biber`? (very low-prio, it works as it is)
3. Make `xetex` work without specifying `dehyph-exptl` and `hyphen-ancientgreek` (is there other way than simply make it dependent on them?)

IMHO after fixing `luatex` we can merge this (and even deprecate the old `texLive`!), all other things seem to work nicely. Actually for me it's already fully usable, as I'm using `xetex`.

**Comment by abbradar** [comment]
Regarding `luatex`: `strace` shows that it tries to run itself... and fails with `-EPERM` (which according to http://man7.org/linux/man-pages/man2/execve.2.html#ERRORS means SUID problem... except we don't have a SUID set nor I have `nosuid` option). `gdb` shows that it's "terminated during startup" which is another thing I've never seen before. I suspect corrupted ELF (maybe `patchelf` or `strip` are the culprits?) but this looks very strange, especially a fail to run itself.

**Comment by vcunat** [comment]
luatex: I suspect it's related to me moving it *after* installation. When I was building all TeX binaries at once, luatex worked fine.

**Comment by abbradar** [comment]
`dontStrip` actually fixed the problem, but it would be interesting to find out why...

**Comment by vcunat** [comment]
Oh, now I remember some interactions between patchelf and strip. One needed to run them in the right order, which I violated by running patchelf by hand too early. I'll try a fix...

**Comment by abbradar** [comment]
One more thing: running `pkgs.texlive.combined.scheme-full` yields
```
these derivations will be built:
  /nix/store/r8bd246skkk7giyqpykbg4r9rs6r4j1k-texlive-combined-2015.drv
building path(s) ‘/nix/store/lsikwd96k6wsgxaby2i85dhvsa4a4z8g-texlive-combined-2015’
while setting up the build environment: executing ‘/nix/store/548wfw1i43glkx8lkyjmbg59h6127qky-bash-4.3-p39/bin/bash’: Argument list too long
builder for ‘/nix/store/r8bd246skkk7giyqpykbg4r9rs6r4j1k-texlive-combined-2015.drv’ failed with exit code 1
error: build of ‘/nix/store/r8bd246skkk7giyqpykbg4r9rs6r4j1k-texlive-combined-2015.drv’ failed
```
I suspect this would be hard to fix...

**Comment by vcunat** [comment]
LuaTeX compiles fine for me now. The problem is described in [patchelf's readme](https://github.com/NixOS/patchelf).

`Argument list too long`: that won't be trivial. This is likely just passing the list of packages to buildEnv, and I *do* deduplicate the list already. Well, people wanting *all* can stick with the old packaging for now ;-)

**Comment by abbradar** [comment]
Nice! I've added debug output to Nix to look at the argument list -- waiting for a big rebuild (caused by something else) to look at it. Maybe we can pass something in the environment...

**Comment by vcunat** [comment]
I thought of taking a few most common collections and changing to one nix package per collection for those, so that would save dozens of tiny packages, as I assume almost all people would take these collections anyway. However, the whole collection has >6.5k packages, so it wouldn't make a significant difference.

**Comment by abbradar** [comment]
According to the debug output we hit the limit per one environment string (`MAX_ARG_STRLEN`) [1]. I expect it's indeed the list of packages what is offending.

[1]: http://man7.org/linux/man-pages/man2/execve.2.html (Limits on size of arguments and environment)

**Comment by datakurre** [comment]
@pikajude Could you find time to try this with stdenvDarwinPure?

I tried the example in [texlive-new](https://github.com/vcunat/nixpkgs/blob/p/tex-space-stdenv/pkgs/tools/typesetting/tex/texlive-new/default.nix) with stdenvDarwinPure + chroot and got:
```
make[1]: Leaving directory '/private/var/folders/y6/2zdynm9x3nv6sq5p6780dwdc0000gn/T/nix-build-texlive-bin-2015.drv-0/texlive-20150521-source/Work/texk/texlive/tl_scripts'
/nix/store/wi3g7v6xq9i2wmlqhkpn1wrz8bi1imks-stdenv-darwin/setup: line 79: wrapProgram: command not found
builder for ‘/nix/store/kvidlpncfjicprhxwpbf1mg998iw3j7f-texlive-bin-2015.drv’ failed with exit code 127
cannot build derivation ‘/nix/store/fxvh86x0yxbxq0bmp43m3m974f84mzqd-texlive-combined-2015.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/zw596kgrijgavr7xjmhb32ql3lqhp2lp-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/zw596kgrijgavr7xjmhb32ql3lqhp2lp-env.drv’ failed
```

**Comment by vcunat** [comment]
@datakurre: that darwin-specific problem should be fixed after 058670188a.

**Comment by datakurre** [comment]
@vcunat Thanks! Next darwin-specific (?) issue:

```
make[1]: Leaving directory '/private/var/folders/y6/2zdynm9x3nv6sq5p6780dwdc0000gn/T/nix-build-texlive-core-big.bin-2015.drv-0/texlive-20150521-source/Work/libs/luajit/include'
mv: missing destination file operand after '/nix/store/642wmnrq2w07y1adcgqpvv1mi4d7fiyk-texlive-core-big.bin-2015-luatex/lib/'
Try 'mv --help' for more information.
builder for ‘/nix/store/7nq4vznyj5y99v7kh7i4x8nzy9s0ssi7-texlive-core-big.bin-2015.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/87p0ys22bvb4z6amgj7pqip36vgib5f8-texlive-combined-2015.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/2rkiva34aw5y3kjjz7yfvnl3f5pwjw44-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/2rkiva34aw5y3kjjz7yfvnl3f5pwjw44-env.drv’ failed
```

And this was the example being built:

```
with import <nixpkgs> {};
stdenv.mkDerivation {
  name = "env";
  buildInputs = [
   (texlive.combine {
      inherit (texlive) scheme-small collection-langkorean algorithms cm-super;
   })
  ];
}
```

**Comment by vcunat** [comment]
Ah, that's trivial :-) I needed to move dynamic libs, but they're not ".so" on darwin... try after efec93e.

**Comment by datakurre** [comment]
@vcunat Thanks again. Now patchelf went peekaboo on darwin:

```
patching script interpreter paths in /nix/store/g4xsn3aavz111g7235hs9dp891dkw9ba-texlive-core-big.bin-2015-xetex
/nix/store/wi3g7v6xq9i2wmlqhkpn1wrz8bi1imks-stdenv-darwin/setup: line 68: patchelf: command not found
/nix/store/wi3g7v6xq9i2wmlqhkpn1wrz8bi1imks-stdenv-darwin/setup: line 68: patchelf: command not found
builder for ‘/nix/store/mjnwl6mwi633i0mwqi9g2pfrii8g44bm-texlive-core-big.bin-2015.drv’ failed with exit code 127
cannot build derivation ‘/nix/store/krd3g3kigjs66ypqjk53x1l9dmsg817c-texlive-combined-2015.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/ncybx8h3vndi9fbcq31ffhx6zc7bcl35-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/ncybx8h3vndi9fbcq31ffhx6zc7bcl35-env.drv’ failed
```

**Comment by vcunat** [comment]
That should be good now. What's next?

**Comment by datakurre** [comment]
:-) Not sure, what this is...

```
post-installation fixup
patching ELF executables and libraries in /nix/store/2xdml8y1bszm95nrp4p982s4c27rvbcw-texlive-core-bi
g.bin-2015
/nix/store/2xdml8y1bszm95nrp4p982s4c27rvbcw-texlive-core-big.bin-2015/bin/ctangle
not an ELF executable
/nix/store/2xdml8y1bszm95nrp4p982s4c27rvbcw-texlive-core-big.bin-2015/bin/ctie
not an ELF executable
/nix/store/2xdml8y1bszm95nrp4p982s4c27rvbcw-texlive-core-big.bin-2015/bin/cweave
...
gzipping man pages in /nix/store/g2if5lm4vx9qi0piq9570q2w6p254lw3-texlive-core-big.bin-2015-luatex
stripping (with flags -S) in /nix/store/g2if5lm4vx9qi0piq9570q2w6p254lw3-texlive-core-big.bin-2015-luatex/lib  /nix/store/g2if5lm4vx9qi0piq9570q2w6p254lw3-texlive-core-big.bin-2015-luatex/bin 
patching script interpreter paths in /nix/store/g2if5lm4vx9qi0piq9570q2w6p254lw3-texlive-core-big.bin-2015-luatex
patching ELF executables and libraries in /nix/store/428yxdwk7ll9jbsl87i0aza9z18dmfvh-texlive-core-big.bin-2015-xetex
/nix/store/428yxdwk7ll9jbsl87i0aza9z18dmfvh-texlive-core-big.bin-2015-xetex/bin/xetex
not an ELF executable
gzipping man pages in /nix/store/428yxdwk7ll9jbsl87i0aza9z18dmfvh-texlive-core-big.bin-2015-xetex
stripping (with flags -S) in /nix/store/428yxdwk7ll9jbsl87i0aza9z18dmfvh-texlive-core-big.bin-2015-xetex/bin 
patching script interpreter paths in /nix/store/428yxdwk7ll9jbsl87i0aza9z18dmfvh-texlive-core-big.bin-2015-xetex
not an ELF executable
not an ELF executable
builder for ‘/nix/store/zgs2qrjlfmzz4xv6hg9m2fmqmsggidn3-texlive-core-big.bin-2015.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/a91zzirydbfvilklnl3gv17qbn9i0nkk-texlive-combined-2015.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/xivjsggw2x5ilq02pmvxzlx04lk7kn9k-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/xivjsggw2x5ilq02pmvxzlx04lk7kn9k-env.drv’ failed
```

**Comment by vcunat** [comment]
I don't see any error. Would you post ```nix-store -l /nix/store/zgs2qrjlfmzz4xv6hg9m2fmqmsggidn3-texlive-core-big.bin-2015.drv | curl -F 'sprunge=<-' http://sprunge.us```

**Comment by datakurre** [comment]
http://sprunge.us/cDYR

**Comment by vcunat** [comment]
Unfortunately, I can't see anything that might cause it to error.

**Comment by datakurre** [comment]
Also never seen failure like that before.

**Comment by datakurre** [comment]
@vcunat Once I removed postFixup for core-big -derivation, the build passed. Something there.

**Comment by vcunat** [comment]
Then the `not an ELF executable` messages are patchelf failing on the `$luatex/bin/*` files. You can have a look what garbage is in there instead of the ELFs.

**Comment by datakurre** [comment]
@vcunat What should I look for and how? readelf gives the same error

**Comment by vcunat** [comment]
Try `file`, for example.

**Comment by datakurre** [comment]
I feel completely clueless.

```
file /nix/store/n4s5r4c1livkgn596j3kaacc7rhwygz2-texlive-core-big.bin-2015-luatex/bin/luatex
/nix/store/n4s5r4c1livkgn596j3kaacc7rhwygz2-texlive-core-big.bin-2015-luatex/bin/luatex: Mach-O 64-bit executable x86_64
```

**Comment by vcunat** [comment]
Oh, I forgot darwin doesn't even use ELF (but Mach-O), so patchelf likely doesn't make sense at all.

**Comment by vcunat** [comment]
@datakurre: I suspect it's enough not to use patchelf at all. When you remove `postFixup`, does `./result-luatex/bin/luatex --version` work?

**Comment by datakurre** [comment]
@vcunat Does not seem to run:

```
result/bin/luatex --version
dyld: Library not loaded: /nix/store/c6rpyb2cn29s1y0aaw0qb1ghaj11zpi9-texlive-core-big.bin-2015/lib/libtexlua52.5.dylib
  Referenced from: /nix/store/li9w6azjbcadg85xh7x8agkp4a182giw-texlive-core-big.bin-2015-luatex/bin/luatex
  Reason: image not found
[1]    80059 trace trap  result/bin/luatex --version
```

**Comment by vcunat** [comment]
Right, this particular problem should also be good now.

**Comment by datakurre** [comment]
@vcunat Thanks a lot. Looks better now:

```
$ result/bin/luatex --version
This is LuaTeX, Version beta-0.80.0 (Web2C 2015/NixOS.org) (rev 5238)

Execute  'luatex --credits'  for credits and version details.

There is NO warranty. Redistribution of this software is covered by
the terms of the GNU General Public License, version 2 or (at your option)
any later version. For more information about these matters, see the file
named COPYING and the LuaTeX source.

Copyright 2015 Taco Hoekwater, the LuaTeX Team.
```

Next I'll try to use this for a real LaTeX project with a good amount of packages (may take a day or a few day) and report back if I find any issues.

**Comment by vcunat** [comment]
It will be nice to have another platform tested :-)

**Comment by datakurre** [comment]
@vcunat I seem to need more hand-holding. How should I debug this?

```
Running 'latex  -recorder  "thesis.tex"'
------------
This is pdfTeX, Version 3.14159265-2.6-1.40.16 (TeX Live 2015/NixOS.org) (preloaded format=.latex-wra
pped)
 restricted \write18 enabled.

kpathsea: Running mktexfmt .latex-wrapped.fmt
mktexfmt: mktexfmt is using the following fmtutil.cnf files (in precedence order):
mktexfmt:   /nix/store/ash5lc6xvrj0gqb5snd9yw714ri984l5-texlive-combined-2015/share/texmf/web2c/fmtut
il.cnf
mktexfmt: mktexfmt is using the following fmtutil.cnf file for writing changes:
mktexfmt:   /Users/datakurre/.texlive2014/texmf-config/web2c/fmtutil.cnf
mktexfmt [INFO]: Disabled formats: 10
mktexfmt [INFO]: Not selected formats: 46
I can't find the format file `.latex-wrapped.fmt'!
```

**Comment by datakurre** [comment]
``fmtutil --all`` : http://sprunge.us/XfIW

**Comment by vcunat** [comment]
The problem is it should never run `mktexfmt` with parameters like `.latex-wrapped.fmt`, and I've never seen it do so. This is empty, right?
```
$ ls -lA /nix/store/ash5lc6xvrj0gqb5snd9yw714ri984l5-texlive-combined-2015/bin/.*
```

**Comment by vcunat** [comment]
Oh, you don't run directly on that branch of mine due to #9044, right?

**Comment by datakurre** [comment]
@vcunat I do run directly against your branch, so that might be an issue. May I still ask the simplest git command to get only your texlive work on top of my master based branch?

(The previous dump with error messages http://sprunge.us/ecXS even it may not matter anymore.)

**Comment by datakurre** [comment]
Something easier available than cherry-picking range of commits?

**Comment by abbradar** [comment]
I suppose you can `git rebase --onto your-branch first-commit-before-@vcunat's-changes` while in his branch.

**Comment by vcunat** [comment]
Yes, but I think my branch directly isn't less likely to work, unless you have those `find -perm` problems.

**Comment by datakurre** [comment]
I assume that it would be enough to apply https://github.com/NixOS/nixpkgs/commit/8f33b8cc93006c97a8f6b7ede20214992ed31fa8 and try again... so, rebuilding...

**Comment by datakurre** [comment]
Picking https://github.com/NixOS/nixpkgs/commit/8f33b8cc93006c97a8f6b7ede20214992ed31fa8 did not seem to be enough (the same issue persists). Something else?

**Comment by vcunat** [comment]
This is empty, right?
```
$ ls -lA /nix/store/ash5lc6xvrj0gqb5snd9yw714ri984l5-texlive-combined-2015/bin/.*
```

**Comment by datakurre** [comment]
@vcunat Yes. (Different hash, beause I picked https://github.com/NixOS/nixpkgs/commit/8f33b8cc93006c97a8f6b7ede20214992ed31fa8 but stil empty.)

**Comment by vcunat** [comment]
Hmm, maybe it's resolved now in fe0eec9b07. I hope poppler won't be missing instead, after removing those old fixups.

**Comment by datakurre** [comment]
@vcunat Thanks. Seems resolved. Now I'm probably just missing some packages and come back if I don't learn how to add them...

**Comment by datakurre** [comment]
@vcunat How should define missing packages to be used with texlive.combine?

**Comment by vcunat** [comment]
I suppose you haven't read the user docs: https://github.com/vcunat/nixpkgs/blob/p/tex-space-stdenv/pkgs/tools/typesetting/tex/texlive-new/default.nix

**Comment by datakurre** [comment]
@vcunat I just don't seem to understand it: What to do when the required package does not exist in generated ``texlive-new/pkgs.nix``?

**Comment by vcunat** [comment]
Which package? I'm fairly confident I pulled the whole TeX Live 2015 in there.

**Comment by datakurre** [comment]
@vcunat I'm currently missing the following:

```
       amssymb
       everyshi
       fontenc
       graphicx
       iflang
       inputenc
       remreset
       texcomp
       texnames
       tikz
```

Previously I've been using
```
    (texLiveAggregationFun { paths = [
      texLive
      texLiveExtra
      texLivePGF
    ];})
```

So, possibly those packages are from ``texLiveExtra`` and ``texLivePGF``. Also, in an another project I'd prefer to have a more recent version of a specific package than was packaged into texLive release and was hoping it to be possible to manually override some packages with more recent versions from CTAN.

**Comment by datakurre** [comment]
@vcunat Sorry! I seem to have messing up concepts by just mapping style files into package names. I'll go through those again...

**Comment by vcunat** [comment]
Yeah, for the prefix of your list, the respective packages are: `amsfonts ms latex graphics oberdiek latex`.

**Comment by datakurre** [comment]
@vcunat All ok for me now. Thanks a lot for your patience. This is such a great improvement.

**Comment by vcunat** [comment]
@abbradar: I believe a3d828e resolved the dehypht-like problems that were still happening in some formats.

**Comment by fread2281** [comment]
Context works with `export TEXMF=$out`

**Comment by maxxk** [comment]
@vcunat Thank you for a great work! 
Is it possible to have a separate derivation for compiled formats? Using different customized texlive in default.nix for a number of tex projects with different packages (e.g. presentations, thesis, papers) but with the same  set of engines results in a lot of texlive-combined-2015 derivations, 8+Mb each.

**Comment by vcunat** [comment]
@fread2281: you used only stuff from inside the `texlive` attribute, right? Can you provide some test case(s).

@maxxk: sounds possible, but a nontrivial thing to implement, I think, so it won't be very soon (from me at least). Possible workarounds for now: create an env that can do all that you need (attribute maps can be unioned easily); or if it currently creates identical files, `nix-store --optimize` should hardlink them (though that's quite a sledge-hammer).

It's funny how a couple weeks ago we had ~1 GB granularity and now we would deal with ~5-20 MB per environment. (The sheer number of files, dirs and symlinks surely has its costs in meta-data as well.)

**Comment by vcunat** [comment]
I'm closing, as the title got implemented. That doesn't mean we can't continue the discussion, though it might be better to open individual issues...

**Comment by fread2281** [comment]
@vcunat I'm using scheme-context. `context --generate` searches `/nix/store` without TEXMF set. However, setting TEXMF breaks bibtex in context (I'm not sure why), and `BSTINPUTS=$TEXMF/share/texmf/bibtex/bst/context` fixes it.

**Comment by nicowilliams-work** [comment]
I'm struggling to figure out how to get [`url.sty`](https://ctan.org/pkg/url) installed.  Or how to use CTAN with Nix.  This issue is the closest thing I've found.  Would a kind soul give me a clue on a tip?

**Comment by 7c6f434c** [comment]
>I'm struggling to figure out how to get [`url.sty`](https://ctan.org/pkg/url) installed.  Or how to use CTAN with Nix.  This issue is the closest thing I've found.  Would a kind soul give me a clue on a tip?

Generally the thing you put into a package list looks like:
```
pkgs.texlive.combine {
  inherit (pkgs.texlive) scheme-small url;
}
```

This expression is a TeXLive package with `url` included.

**Comment by StefanSchroeder** [comment]
I have looked in multiple places but it seems to me that the use of TeX within Nix/Nixos is not very well documented. While there is a (short) Wiki entry (https://nixos.wiki/wiki/TexLive), there are a couple of re-occurring themes that show up in the issues again and again:

- OP wants to use a self-made (self-hosted) package (=tex-package/sty) 
- OP wants to use a package from Texlive (and probably on CTAN) that is not in nixpkgs
- OP wants to use a package that is on CTAN, but not in Texlive

In the forums the typical response to these types of request is: Here is a link to the derivation that I made for this other package, you can adapt it to your use-case. While this probably does work, it's a significant burden, therefore it seems worthwhile to 

- document the usage of CTAN packages generically or
- provide wrapper functions for CTAN-package inclusion.

I am not proficient in writing these functions, but it looks a sufficiently good idea to me that I kindly request it from the Texlive maintainers of nixpkgs.

@veprbl

**Comment by veprbl** [comment]
@StefanSchroeder 
In your message you've never mentioned the nixpkgs manual. Have you seen this section:
https://nixos.org/manual/nixpkgs/stable/#sec-language-texlive-custom-packages
Is there something that you find missing there?

Regarding the generic wrapper, I would rather discuss it in a separate issue. And also I would like to mention that IIRC there weren't any non-texlive packages submitted to nixpkgs. Before we look into implementing any generic CTAN wrappers, it would be nice to see some concrete use cases.

**Comment by StefanSchroeder** [comment]
Oh, I have of course consulted the manual. As a new Nixos-user I find the relevant section perplexingly complicated. 

Let me lay out the use-cases that I personally have. As a long-time TeX-user I have a accumulated a variety of documents that use various packages. My ability to process them is like a test-bed to see if Nixos is for me.

E.g.
https://ctan.org/pkg/memoir (which is in TL)
https://ctan.org/pkg/acrotex (which is not in TL)

Although _memoir_ is contained in TL, it cannot be found using https://search.nixos.org/packages?query=memoir.
Thus it remains unclear what to do to install the memoir-package. I think installing Texlive-full, just to be sure is not a nice solution.

The package _acrotex_ is not in TL (although the license seems to be a free one, I guess: The LaTeX Project Public License).
I have no clue how to adapt the recipe in the manual (15.22.2)  to install it. When including the package in Miktex, the necessary dependencies are downloaded and installed automatically. I wonder if the Miktex-package format could be used cross-platform  (for non-binary-package).

That being said, I was wondering where the transition from TL-2019 to TL-2020 is being coordinated. I found  comments in  https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/typesetting/tex/texlive/default.nix that seem to indicate that people are working on it, but I can't see what's necessary...

**Comment by Mathnerd314** [comment]
This issue is pretty old, I would suggest opening a new issue for your use-case.

**Comment by StefanSchroeder** [comment]
OK, will do.
